src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -106,32 +106,65 @@ std::shared_ptr<RayObject> GetRequest::Get(const ObjectID &object_id) const {
 
 CoreWorkerMemoryStore::CoreWorkerMemoryStore() {}
 
-Status CoreWorkerMemoryStore::Put(const ObjectID &object_id, const RayObject &object) {
-  std::unique_lock<std::mutex> lock(lock_);
-  auto iter = objects_.find(object_id);
-  if (iter != objects_.end()) {
-    return Status::ObjectExists("object already exists in the memory store");
+void CoreWorkerMemoryStore::GetAsync(
+    const ObjectID &object_id, std::function<void(std::shared_ptr<RayObject>)> callback) {
+  std::shared_ptr<RayObject> ptr;
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    auto iter = objects_.find(object_id);
+    if (iter != objects_.end()) {
+      ptr = iter->second;
+    } else {
+      object_async_get_requests_[object_id].push_back(callback);
+    }
+  }
+  // call outside lock
+  if (ptr != nullptr) {
+    callback(ptr);
   }
+}
 
+Status CoreWorkerMemoryStore::Put(const ObjectID &object_id, const RayObject &object) {
+  std::vector<std::function<void(std::shared_ptr<RayObject>)>> async_callbacks;
   auto object_entry =
       std::make_shared<RayObject>(object.GetData(), object.GetMetadata(), true);
 
-  bool should_add_entry = true;
-  auto object_request_iter = object_get_requests_.find(object_id);
-  if (object_request_iter != object_get_requests_.end()) {
-    auto &get_requests = object_request_iter->second;
-    for (auto &get_request : get_requests) {
-      get_request->Set(object_id, object_entry);
-      if (get_request->ShouldRemoveObjects()) {
-        should_add_entry = false;
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    auto iter = objects_.find(object_id);
+    if (iter != objects_.end()) {
+      return Status::ObjectExists("object already exists in the memory store");
+    }
+
+    bool should_add_entry = true;
+    auto object_request_iter = object_get_requests_.find(object_id);
+    if (object_request_iter != object_get_requests_.end()) {
+      auto &get_requests = object_request_iter->second;
+      for (auto &get_request : get_requests) {
+        get_request->Set(object_id, object_entry);
+        if (get_request->ShouldRemoveObjects()) {
+          should_add_entry = false;
+        }
       }
     }
+
+    auto async_callback_it = object_async_get_requests_.find(object_id);
```

Comment:
Can you move this block after the `if should_add_entry` one? The order is strange right now.

Reply:
Done